File: src/step2point/algorithms/merge_within_cell.py
from __future__ import annotations

import numpy as np

from step2point.algorithms.base import CompressionAlgorithm
from step2point.core.results import CompressionResult
from step2point.core.shower import Shower
from step2point.cpp_backend import cpp_available, merge_within_cell as cpp_merge_within_cell


class MergeWithinCell(CompressionAlgorithm):
    """Merging within the detector cell.

    All steps that fall within the same detector cell (have the same cell ID)
    are merged, no matter what are the other properties.

    TODO Implement a time window to merge only steps within, and ignore the rest.
    """
    name = "merge_within_cell"

    def __init__(self, backend: str = "auto") -> None:
        if backend not in {"auto", "python", "cpp"}:
            raise ValueError("backend must be 'auto', 'python', or 'cpp'.")
        self.backend = backend

    def compress(self, shower: Shower) -> CompressionResult:
        if self.backend == "cpp":
            return cpp_merge_within_cell(shower)
        if self.backend == "auto" and cpp_available():
            return cpp_merge_within_cell(shower)
        return self._compress_python(shower)

    def _compress_python(self, shower: Shower) -> CompressionResult:
        if shower.cell_id is None:
            raise ValueError("MergeWithinCell requires cell_id.")
        unique_cells, inverse = np.unique(shower.cell_id, return_inverse=True)
        n = len(unique_cells)
        e_sum = np.bincount(inverse, weights=shower.E, minlength=n)
        safe_e = np.where(e_sum > 0.0, e_sum, 1.0)
        x = np.bincount(inverse, weights=shower.x * shower.E, minlength=n) / safe_e
        y = np.bincount(inverse, weights=shower.y * shower.E, minlength=n) / safe_e
        z = np.bincount(inverse, weights=shower.z * shower.E, minlength=n) / safe_e
        t = None
        if shower.t is not None:
            t = np.bincount(inverse, weights=shower.t * shower.E, minlength=n) / safe_e
        out = Shower(
            shower_id=shower.shower_id,
            x=x.astype(np.float32),
            y=y.astype(np.float32),
            z=z.astype(np.float32),
            E=e_sum.astype(np.float32),
            t=None if t is None else t.astype(np.float32),
            cell_id=unique_cells.astype(np.uint64),
            primary=shower.primary,
            metadata={**shower.metadata, "algorithm": self.name, "backend": "python"},
        )
        return CompressionResult(
            shower=out,
            algorithm=self.name,
            stats={
                "n_points_before": shower.n_points,
                "n_points_after": out.n_points,
                "compression_ratio": out.n_points / max(shower.n_points, 1),
                "energy_before": shower.total_energy,
                "energy_after": out.total_energy,
            },
        )

File: cpp/src/merge_within_cell.cpp
#include "step2point/algorithms.hpp"

#include <algorithm>
#include <stdexcept>
#include <unordered_map>
#include <vector>

namespace step2point {

CompressionResult merge_within_cell(const Shower& input) {
  if (!input.valid()) {
    throw std::runtime_error("Invalid Shower passed to merge_within_cell().");
  }
  if (!input.cell_id) {
    throw std::runtime_error("merge_within_cell() requires cell_id.");
  }

  struct Acc {
    double e = 0.0;
    double x = 0.0;
    double y = 0.0;
    double z = 0.0;
    double t = 0.0;
  };

  std::unordered_map<std::uint64_t, Acc> accum;
  accum.reserve(input.n_points());
  std::vector<std::uint64_t> cell_order;
  cell_order.reserve(input.n_points());

  for (std::size_t i = 0; i < input.n_points(); ++i) {
    const auto cell = (*input.cell_id)[i];
    auto [it, inserted] = accum.try_emplace(cell, Acc{});
    if (inserted) {
      cell_order.push_back(cell);
    }
    auto& a = it->second;
    const double e = input.E[i];
    a.e += e;
    a.x += input.x[i] * e;
    a.y += input.y[i] * e;
    a.z += input.z[i] * e;
    if (input.t) a.t += (*input.t)[i] * e;
  }

  std::sort(cell_order.begin(), cell_order.end());

  Shower out;
  out.shower_id = input.shower_id;
  out.cell_id = std::vector<std::uint64_t>{};
  if (input.t) out.t = std::vector<float>{};

  out.x.reserve(cell_order.size());
  out.y.reserve(cell_order.size());
  out.z.reserve(cell_order.size());
  out.E.reserve(cell_order.size());
  out.cell_id->reserve(cell_order.size());
  if (out.t) out.t->reserve(cell_order.size());

  for (const auto cell : cell_order) {
    const auto& a = accum.at(cell);
    const double safe = a.e > 0.0 ? a.e : 1.0;
    out.cell_id->push_back(cell);
    out.x.push_back(static_cast<float>(a.x / safe));
    out.y.push_back(static_cast<float>(a.y / safe));
    out.z.push_back(static_cast<float>(a.z / safe));
    out.E.push_back(static_cast<float>(a.e));
    if (out.t) out.t->push_back(static_cast<float>(a.t / safe));
  }

  CompressionResult result;
  result.shower = std::move(out);
  result.algorithm = "merge_within_cell";
  result.stats.n_points_before = input.n_points();
  result.stats.n_points_after = result.shower.n_points();
  result.stats.energy_before = input.total_energy();
  result.stats.energy_after = result.shower.total_energy();
  result.stats.compression_ratio = input.n_points() == 0
      ? 1.0
      : static_cast<double>(result.shower.n_points()) / static_cast<double>(input.n_points());
  return result;
}

} // namespace step2point
